### DISCUSSION TITLE
Creating a Proxy Server with Fast API

### DISCUSSION THREAD

**Discussion by mishrakeshav** [fastapi/fastapi#9599] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from starlette.requests import Request
from starlette.responses import StreamingResponse
from starlette.background import BackgroundTask

import httpx

client = httpx.AsyncClient(base_url="http://containername:7800/")

async def _reverse_proxy(request: Request):
    url = httpx.URL(path=request.url.path,
                    query=request.url.query.encode("utf-8"))
    rp_req = client.build_request(request.method, url,
                                  headers=request.headers.raw,
                                  content=await request.body())
    rp_resp = await client.send(rp_req, stream=True)
    return StreamingResponse(
        rp_resp.aiter_raw(),
        status_code=rp_resp.status_code,
        headers=rp_resp.headers,
        background=BackgroundTask(rp_resp.aclose),
    )

app.add_route("/titles/{path:path}",
              _reverse_proxy, ["GET", "POST"])
```

### Description

I want to create a proxy server using Fast API which add few extra headers and re-routes them a different url and returns the response. 
I have refered https://github.com/tiangolo/fastapi/issues/1788 but this would be possible only if they are on the same machine. I want to re-route the request to different urls. 
I'm trying to find out the best method to do this. Please suggest any possible solution(s) 

### Operating System

Linux

### FastAPI Version

0.95

### Python Version

3.7

**Comment by jgould22** [comment]
>   this would be possible only if they are on the same machine

Why is that the case?

**Reply by mishrakeshav** [reply]
I tried it for the following url:  https://www.google.com/
I was getting 404 error. 
Tried the same code for another server hosted on a diffferent machine with a dns, getting the same 404 response. 

Tried it for a server running on the same machine, getting proper responses. 
ALso, in the example they has mentioned containername in the url

**Comment by BilalAlpaslan** [comment]
I think some header fields cause that.

**Reply by BilalAlpaslan** [reply]
you can try delete or fix host and accept-encoding fields in header

```py
from fastapi import FastAPI, Request
from fastapi.responses import StreamingResponse
import requests

app = FastAPI(debug=True)

base_url = "https://www.google.com/"


@app.get("/titles/{path:path}")
async def read_root(request: Request, path: str):
    headers = {k: v for k, v in request.headers.items()}
    headers.pop('host')
    headers.pop('accept-encoding')

    res = requests.get(f'{base_url}{path}', params=request.query_params, headers=headers, stream=True)

    return StreamingResponse(
        res.iter_lines(),
        status_code=res.status_code,
        headers=headers,
    )

if __name__ == "__main__":
    import uvicorn
    uvicorn.run('st:app', reload=True)

```

**Comment by alihssan** [comment]
**Endpoint Mapping:**
A dictionary named ENDPOINT_MAPPING is defined to map specific paths to their corresponding target hosts.
Example: "/api" maps to "http://api.example.com", "/static" maps to "http://static.example.com", etc.

**Reverse Proxy Middleware:**
A custom middleware class named ReverseProxyMiddleware is created by extending BaseHTTPMiddleware.
The dispatch method of the middleware is overridden to handle the reverse proxy logic.
The middleware iterates through the items in ENDPOINT_MAPPING to find a matching path.
If a match is found, it constructs the target URL, forwards the request to the target server, and returns the response.
If no match is found, it proceeds with the normal routing by calling call_next(request).

**Application Setup:**
The FastAPI application (app) is created.
The target server URL (TARGET_SERVER) is defined as "http://example.com".
The **ReverseProxyMiddleware** is added to the app using app.add_middleware(ReverseProxyMiddleware).

```bash
from fastapi import FastAPI, HTTPException
from fastapi.middleware.base import BaseHTTPMiddleware
import httpx

app = FastAPI()

# Define the target server URL
TARGET_SERVER = "http://example.com"

# Endpoint mapping for reverse proxy
ENDPOINT_MAPPING = {
    "/api": "http://api.example.com",
    "/static": "http://static.example.com",
}

class ReverseProxyMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        # Check if the path matches an endpoint in the mapping
        for path, target_host in ENDPOINT_MAPPING.items():
            if request.url.path.startswith(path):
                # Forward the request to the target server
                target_url = f"{target_host}{request.url.path}"
                async with httpx.AsyncClient() as client:
                    response = await client.request(request.method, target_url, headers=dict(request.headers), data=await request.body())

                # Return the response to the client
                return response

        # If no matching endpoint found, proceed with the normal routing
        response = await call_next(request)
        return response

# Add reverse proxy middleware to the app
app.add_middleware(ReverseProxyMiddleware)

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="127.0.0.1", port=8000)
```

**Reply by leeprevost** [reply]
I love this approach.  Is this now the recommended way to handle multiple reverse proxy routes in FASTApi?   Or has this evolved?   One question is how to handle multiple proxies for ports on the same server to just avoid CORs issues.

**Comment by shashstormer** [comment]
you can try something like https://github.com/shashstormer/m3u8_proxy-cors

**Comment by asmaier** [comment]
Here is a simple example for re-routing GET and POST requests to `https://example.com/api`:

```
from fastapi import Request, Response
import httpx

URL = "https://example.com/api"

@router.get("/{sub_url:path}")
@router.post("/{sub_url:path}")
async def proxy_request(sub_url: str, request: Request, response: Response):
    content = request.stream()
    headers = dict(request.headers)
    method = request.method
    url = URL + sub_url

    # We need to remove the "host" header, otherwise the proxied request will fail
    headers.pop("host", None)

    async with httpx.AsyncClient() as client:
        proxy = await client.request(method, url, timeout=60, headers=headers, content=content)
        response.body = proxy.content
        response.status_code = proxy.status_code

    return response
```

**Comment by phungtcb99** [comment]
I think using middleware is enough for your usecases
